src/v/cloud_storage/access_time_tracker.cc

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/access_time_tracker.h"
+
+#include "serde/serde.h"
+#include "units.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/smp.hh>
+
+#include <absl/container/btree_map.h>
+
+#include <exception>
+#include <variant>
+
+namespace absl {
+
+template<class Key, class Value>
+void write(iobuf& out, const btree_map<Key, Value>& btree) {
+    using serde::write;
+    write(out, static_cast<uint64_t>(btree.size()));
+    for (auto it : btree) {
+        write(out, it.first);
+        write(out, it.second);
+    }
+}
+
+template<class Key, class Value>
+void read_nested(
+  iobuf_parser& in,
+  btree_map<Key, Value>& btree,
+  size_t const bytes_left_limit) {
+    using serde::read_nested;
+    uint64_t sz;
+    read_nested(in, sz, bytes_left_limit);
+    for (auto i = 0UL; i < sz; i++) {
+        Key key;
+        Value value;
+        read_nested(in, key, bytes_left_limit);
+        read_nested(in, value, bytes_left_limit);
+        btree.insert({key, value});
+    }
+}
+
+} // namespace absl
+
+namespace cloud_storage {
+
+void access_time_tracker::add_timestamp(
+  std::string_view key, std::chrono::system_clock::time_point ts) {
+    uint32_t seconds = std::chrono::time_point_cast<std::chrono::seconds>(ts)
+                         .time_since_epoch()
+                         .count();
+    uint32_t hash = xxhash_32(key.data(), key.size());
+    _table.data[hash] = seconds;
+    _dirty = true;
+}
+
+void access_time_tracker::remove_timestamp(std::string_view key) noexcept {
+    try {
+        uint32_t hash = xxhash_32(key.data(), key.size());
+        _table.data.erase(hash);
```

Comment:
Interesting. Since this is a count-min sketch, we don't really support delete, right? IIUC, any other cache items which map to same key (hash collision) will become "oldest access time" when we delete the other element. I wonder if there are other approaches here (e.g. set timestamp to curr_time - X or some other heuristic to age (make older) the entry? Might also consider a different cache replacement algorithm in the future (maybe CLOCK or CAR)?

Reply:
That makes more sense. It is very similar to a k=1 CM Sketch, though. 🙂  My understanding was that CM Sketch has a fixed size that doesn't grow. I think this is the "sublinear" benefit you get from a probabilistic datastructure. Anyways, I think you have a good API here, which will allow us to change the implementation in the future. The fact that this data is not required for correctness (only performance, right?) might make on-disk format upgrades easier--just delete the old version on upgrade and create the new one on demand (initially, all segments will be the same age).

Reply:
yep, it's only for performance so we're free to change it
CM sketch approach worked fine, unless you have terrabytes of cache. In this case you will use extra MiB or so for every TiB in cache. But CM sketch have to be tuned for worst case scenario (a lot of small segments). If the actual segments are not small we're wasting memory. But this approach is adaptive in this regard. It will do well in both cases (a lot of small segment or fewer large segments).